Using the library from C++17 after building libraries in C++11 mode (Arch)

The discussion started in Arch bug tracker at https://bugs.archlinux.org/task/65953 and I'd like to ask for opinion of fixing the issue here.

In Arch pre-built tensorflow_cc binaries are built with C++11, and that causes problems when using them from C++17.

Tensorflow_cc library uses own copy of Absl, and uses `absl::string_view` in function signatures.
`absl::string_view` is mapped to `std::string_view` if C++17 is detected, and to own implementation if C++17 is not.
That leads to linker errors like this when using Arch tensorflow_cc library from C++17 code:
```
test.cc:(.text+0x2b6): undefined reference to `tensorflow::ops::Conv2D::Conv2D(tensorflow::Scope const&, tensorflow::Input, tensorflow::Input, absl::Span<int const> const&, std::basic_string_view<char, std::char_traits<char> >)'
```

One workaround is to comment out this line in /usr/include/tensorflow/absl/base/config.h:
```
#define ABSL_HAVE_STD_STRING_VIEW 1
```

This will make the library look for the custom absl::string_view implementation even if tensorflow is called from C++17.

It doesn't look nice, so I wonder if you have any other suggestion?

--- a/tensorflow_cc/CMakeLists.txt
+++ b/tensorflow_cc/CMakeLists.txt
@@ -11,7 +11,7 @@ option(REQUIRE_CUDA "Make sure to find and use CUDA (implies ALLOW_CUDA)." OFF)
 set(LOCAL_RAM_RESOURCES 4096 CACHE STRING "The amount of local RAM resources passed to bazel scheduler (e.g., 4096).")
 set(LOCAL_CPU_RESOURCES HOST_CPUS CACHE STRING "The amount of local CPU cores passed to bazel scheduler (e.g., 2).")
 set(TENSORFLOW_TAG "v${version}" CACHE STRING "The tensorflow release tag to be checked out (default v${version}).")
-set(CMAKE_CXX_STANDARD 17 CACHE STRING "The C++ standard for building and linking the library (e.g., 17).")
+set(CMAKE_CXX_STANDARD 14 CACHE STRING "The C++ standard for building and linking the library (e.g., 14).")
 
 # -------------
 # CMake Options

--- a/tensorflow_cc/CMakeLists.txt
+++ b/tensorflow_cc/CMakeLists.txt
@@ -5,13 +5,12 @@ project(
 )
 set(PROJECT_VERSION "${version}")
 
-# If enabled, bazel has to be installed.
 option(ALLOW_CUDA "Try to find and use CUDA." ON)
 option(REQUIRE_CUDA "Make sure to find and use CUDA (implies ALLOW_CUDA)." OFF)
 set(LOCAL_RAM_RESOURCES 4096 CACHE STRING "The amount of local RAM resources passed to bazel scheduler (e.g., 4096).")
 set(LOCAL_CPU_RESOURCES HOST_CPUS CACHE STRING "The amount of local CPU cores passed to bazel scheduler (e.g., 2).")
 set(TENSORFLOW_TAG "v${version}" CACHE STRING "The tensorflow release tag to be checked out (default v${version}).")
-set(TARGET_CXX_STANDARD "cxx_std_11" CACHE STRING "C++ standard to be enforced when linking to TensorflowCC targets (e.g., cxx_std_11).")
+set(CMAKE_CXX_STANDARD 17 CACHE STRING "The C++ standard for building and linking the library (e.g., 17).")
 
 # -------------
 # CMake Options
@@ -22,9 +21,6 @@ set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/lib")
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/lib")
 include(CMakePackageConfigHelpers)
 set(CMAKECFG_INSTALL_DIR lib/cmake/TensorflowCC)
-
-set_property(CACHE TARGET_CXX_STANDARD PROPERTY STRINGS
-    "cxx_std_11" "cxx_std_14" "cxx_std_17" "cxx_std_20")
 
 # Configure the build_tensorflow script.
 configure_file("cmake/build_tensorflow.sh.in" "build_tensorflow.sh" @ONLY)
@@ -40,7 +36,7 @@ include(TensorflowBase)
 # ------------------------------
 
 add_library(tensorflow_cc INTERFACE)
-target_compile_features(tensorflow_cc INTERFACE ${TARGET_CXX_STANDARD})
+target_compile_features(tensorflow_cc INTERFACE "cxx_std_${CMAKE_CXX_STANDARD}")
 
 # The include folders are sometimes contained under bazel-bin/bin/ and sometimes just bazel-bin.
 target_include_directories(

--- a/tensorflow_cc/cmake/build_tensorflow.sh.in
+++ b/tensorflow_cc/cmake/build_tensorflow.sh.in
@@ -95,6 +95,13 @@ else
     export TF_NEED_CUDA=0
 fi
 
+# set bazel options for c++17 standard if necessary
+cxx_std_opts=""
+if [ "@CMAKE_CXX_STANDARD@" == 17 ]; then
+    echo "Using C++17 standard"
+    cxx_std_opts="--config=c++17_gcc"
+fi
+
 # configure and build
 ./configure
 JVM_RAM_RESOURCES=$((@LOCAL_RAM_RESOURCES@ / 4))
@@ -105,6 +112,7 @@ bazel --host_jvm_args="-Xmx${JVM_RAM_RESOURCES}m" \
             --local_ram_resources=@LOCAL_RAM_RESOURCES@ \
             --local_cpu_resources=@LOCAL_CPU_RESOURCES@ \
             --discard_analysis_cache \
+            $cxx_std_opts \
             $cuda_config_opts \
             tensorflow:libtensorflow_cc.so \
             tensorflow:install_headers

